src/main/java/fi/aalto/cs/apluscourses/ui/base/CompositeTreeNode.java

```diff
@@ -0,0 +1,79 @@
+package fi.aalto.cs.apluscourses.ui.base;
+
+import fi.aalto.cs.apluscourses.presentation.base.FilterableTree;
+import java.util.Collections;
+import java.util.Enumeration;
+import java.util.List;
+import java.util.stream.Collectors;
+import javax.swing.tree.TreeNode;
+import jdk.internal.jline.internal.Nullable;
+import org.jetbrains.annotations.NotNull;
+
+public class CompositeTreeNode implements TreeNode {
```

Comment:
This implementation looks good, but now that there is no longer the additional visibility property, I wonder if we need this class at all? Since it's only used internally in `TreeView`, we could use `DefaultMutableTreeNode` there to avoid reinventing the wheel, and the mutability wouldn't matter, since the mutable tree node is just an interal implementation detail of `TreeView`.

Reply:
That's sounds reasonable enough, even though this class is so basic it barely brings any complexity.

I'd like to keep the tree construction logic out of `TreeView` class for better unit-testability but that shouldn't be a problem; a util class or something will do.